server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1328,6 +1344,44 @@ final ReaderContext createAndPutReaderContext(
         }
     }
 
+    final ReaderContext createAndPutRelocatedPitContext(
+        ShardSearchContextId contextId,
+        IndexService indexService,
+        IndexShard shard,
+        Engine.SearcherSupplier reader,
+        long keepAliveInMillis
+    ) {
+        ReaderContext readerContext = null;
+        try {
+            long newKey = idGenerator.incrementAndGet();
+            // Check that we don't already have a relocation mapping for this context id
+            final Long previous = activeReaders.generateRelocationMapping(contextId, newKey);
+            if (previous == null) {
+                readerContext = new ReaderContext(contextId, indexService, shard, reader, keepAliveInMillis, false);
+                reader = null;
+                final ReaderContext finalReaderContext = readerContext;
+                final SearchOperationListener searchOperationListener = shard.getSearchOperationListener();
+                searchOperationListener.onNewReaderContext(finalReaderContext);
+                readerContext.addOnClose(() -> searchOperationListener.onFreeReaderContext(finalReaderContext));
+                activeReaders.putRelocatedReader(newKey, readerContext);
+                // ensure that if we race against afterIndexRemoved, we remove the context from the active list.
+                // this is important to ensure store can be cleaned up, in particular if the search is a scroll with a long timeout.
+                final Index index = readerContext.indexShard().shardId().getIndex();
+                if (indicesService.hasIndex(index) == false) {
+                    removeReaderContext(readerContext.id());
+                    throw new IndexNotFoundException(index);
+                }
+                readerContext = null;
+                return finalReaderContext;
+            } else {
+                // we already have a mapping for this context, dont add a new one and use the existing instead
```

Comment:
We might encounter a race condition between two retries: one tries to put the relocation key first, while the second detects that the relocation key already exists and tries to get the relocation reader. However, the first retry may not have added the reader to the map yet - only added the relocation key. This isn't a blocker, but it's something to be aware of.